pkg/docker/config/config.go

```diff
@@ -0,0 +1,281 @@
+	"github.com/docker/docker/pkg/homedir"
+	"github.com/pkg/errors"
+)
+
+type dockerAuthConfig struct {
+	Auth string `json:"auth,omitempty"`
+}
+
+type dockerConfigFile struct {
+	AuthConfigs map[string]dockerAuthConfig `json:"auths"`
+	CredHelpers map[string]string           `json:"credHelpers,omitempty"`
+}
+
+const (
+	dockerHostname = "docker.io"
+	dockerRegistry = "registry-1.docker.io"
+
+	authCfg           = ".config/containers"
+	authCfgFileName   = "auth.json"
+	dockerCfg         = ".docker"
+	dockerCfgFileName = "config.json"
+
+	addAuth    = "add"
+	deleteAuth = "delete"
+)
+
+var (
+	// ErrNotLoggedIn is returned for users not logged into a registry
+	// that they are trying to logout of
+	ErrNotLoggedIn = errors.Errorf("not logged in")
+)
+
+// SetAuthentication stores the username and password in the auth.json file
+func SetAuthentication(registry, username, password string) error {
+	registry = DockerHostNameToRegistry(registry)
+	path, err := getPathToAuth()
+	if err != nil {
+		return err
+	}
+	return modifyJSON(registry, username, password, path, addAuth)
+}
+
+// GetAuthentication returns the registry credentials stored in
+// either auth.json file or .docker/config.json
+func GetAuthentication(registry string) (string, string, error) {
+	registry = DockerHostNameToRegistry(registry)
+	// Check auth.json file first
+	path, err := getPathToAuth()
+	if err != nil {
+		return "", "", errors.Wrapf(err, "error getting path to auth file")
+	}
+	username, password, err := findAuthentication(registry, path)
+	if err != nil {
+		return "", "", err
+	}
+	if username != "" && password != "" {
+		return username, password, nil
+	}
+
+	// check ~/.docker/config.json as secondary option
+	path = filepath.Join(homedir.Get(), dockerCfg, dockerCfgFileName)
+	username, password, err = findAuthentication(registry, path)
+	if err != nil {
+		return "", "", err
+	}
+	return username, password, nil
+}
+
+// RemoveAuthentication deletes the credentials stored in auth.json
+func RemoveAuthentication(registry string) error {
+	registry = DockerHostNameToRegistry(registry)
+	path, err := getPathToAuth()
+	if err != nil {
+		return err
+	}
+	return modifyJSON(registry, "", "", path, deleteAuth)
+}
+
+// DockerHostNameToRegistry converts the docker hostname to the docker registry
+func DockerHostNameToRegistry(registry string) string {
+	if registry == dockerHostname {
+		return dockerRegistry
+	}
+	return registry
+}
+
+// getPath gets the path of the auth.json file since the user can set their own path
+// by setting CRIO_AUTH_PATH
+// If CRIO_AUTH_PATH is not set, the default path is ~/.config/containers/auth.json
+func getPathToAuth() (string, error) {
+	if os.Getenv("CRIO_AUTH_PATH") != "" {
+		return filepath.Join(os.Getenv("CRIO_AUTH_PATH"), authCfgFileName), nil
+	}
+	dir := filepath.Join(homedir.Get(), authCfg)
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		if err = os.Mkdir(dir, 0700); err != nil {
```

Comment:
Same as the discussion around `readJSONFile`: the read path should not be modifying the filesystem at all.

Reply:
Would it be better if I renamed the function? I am not sure where else to check for the existence of the directory and create it if it doesn't exist.

Reply:
I think the `updated` case of `modifyJSON` would be a good place. If it costs an extra `filepath.Dir` call, that seems acceptable.

Reply:
hmm, is it possible to create this directory when crio/kpod is installed? I think that might be a better option 🤔

Reply:
fixed

Reply:
This directory is per user, so no way to create it in install time.  Has to be created during first use.